### PR TITLE
[GPU] Support weight tag for oneDNN v2.6

### PR DESCRIPTION
**Pull request by hyunback**
multiple networks runs failure because of unsupported weight tag.

is_os_xy_isa8_osv8_isv2
os_is_xy_isa8_osv8_isv2
g_os_is_yx_isa8_osv8_isv2
is_os_yx_osa4_isa8_osv8_isv4

Add to check dims stride logic for dims order.

Signed-off-by: hyunback <hyunback.kim@intel.com>

### Tickets:
 - *80685*

### CODE REVIEW

**Comment by isanghao** [comment]
@hyunback  Please update onednn submodule, too.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 1]
```diff
@@ -313,18 +313,50 @@ cldnn::format find_format(dnnl::memory::desc desc, bool is_grouped) {
     if (onednn_desc != dnnl::memory::format_tag::undef) {
         return convert_format(onednn_desc, is_grouped);
     } else {
+        auto blk = desc.data.format_desc.blocking;
+
+        auto strides = desc.data.format_desc.blocking.strides;
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:318]:
```suggestion
        auto strides = blk.strides;
```

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 2]
```diff
@@ -313,18 +313,50 @@ cldnn::format find_format(dnnl::memory::desc desc, bool is_grouped) {
...
+        std::vector<size_t> order(desc.data.ndims);
+        std::iota(order.begin(), order.end(), 0);
+        std::sort(order.begin(), order.end(),
+                [&strides] (size_t ind_l, size_t ind_r) {
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:322]: nit: alignment

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 3]
```diff
@@ -313,18 +313,50 @@ cldnn::format find_format(dnnl::memory::desc desc, bool is_grouped) {
...
+                    return strides[ind_l] > strides[ind_r];
+                });
+
         if (is_grouped) {
-            throw std::runtime_error(std::string("Unsupported grouped onednn dnnl::memory::desc find_format"));
+            if (desc.data.ndims == 5 && desc.data.format_desc.blocking.inner_nblks == 3
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:327]:
```suggestion
            if (desc.data.ndims == 5 && blk.inner_nblks == 3
```

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 4]
```diff
@@ -313,18 +313,50 @@ cldnn::format find_format(dnnl::memory::desc desc, bool is_grouped) {
...
+                && blk.inner_idxs[0] == 1 && blk.inner_idxs[1] == 0 && blk.inner_idxs[2] == 1 && blk.inner_idxs[3] == 0
+                && order[0] == 1 && order[1] == 0 && order[2] == 2 && order[3] == 3) {
                 return cldnn::format::is_os_yx_isa2_osa8_isv8_osv2;
-            } else if (desc.data.ndims == 4 && desc.data.format_desc.blocking.inner_nblks == 2 &&
-                blk.inner_blks[0] == 16 && blk.inner_blks[1] == 4 && blk.inner_idxs[0] == 0 && blk.inner_idxs[1] == 1) {
+            } else if (desc.data.ndims == 4 && desc.data.format_desc.blocking.inner_nblks == 2
+                && blk.inner_blks[0] == 16 && blk.inner_blks[1] == 4 && blk.inner_idxs[0] == 0 && blk.inner_idxs[1] == 1
+                && order[0] == 0 && order[1] == 1 && order[2] == 2 && order[3] == 3) {
                 return cldnn::format::os_is_yx_osv16_isv4;
+            } else if (desc.data.ndims == 4 && desc.data.format_desc.blocking.inner_nblks == 3
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:350]:
```suggestion
            } else if (desc.data.ndims == 4 && blk.inner_nblks == 3
```

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 5]
```diff
@@ -419,6 +419,78 @@ inline uint get_giy_xs_os_xsv2_osv_index(uint g, uint o, uint i, uint y, uint x,
         CAT(prefix, _OFFSET),                                                   \
         sub_group_size)
 
+inline uint get_is_os_yx_isa8_osv8_isv2_index(uint o, uint i, uint y, uint x, uint size_x,
+                                                uint size_y, uint size_ifm, uint size_ofm, uint offset)
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:423]: nit: alignment

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 6]
```diff
@@ -419,6 +419,78 @@ inline uint get_giy_xs_os_xsv2_osv_index(uint g, uint o, uint i, uint y, uint x,
...
+{
+	const uint isv2_idx = i % 2;
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:425]: Is it really necessary to use such specific code alignment style?

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 7]
```diff
@@ -419,6 +419,78 @@ inline uint get_giy_xs_os_xsv2_osv_index(uint g, uint o, uint i, uint y, uint x,
...
+
+    return idx;
+}
+
+inline uint get_g_os_is_yx_isa8_osv8_isv2_index(uint g, uint o, uint i, uint y, uint x, uint size_x,
+                                                uint size_y, uint size_ifm, uint size_ofm, uint offset)
+{
+	const uint isv2_idx = i % 2;
+	const uint osv_idx = o % 8;
+	const uint isv1_idx = (i / 2) % 8;
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:450]: And the same question is here.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 8]
```diff
@@ -419,6 +419,78 @@ inline uint get_giy_xs_os_xsv2_osv_index(uint g, uint o, uint i, uint y, uint x,
...
+    return idx;
+}
+
+#define GET_FILTER_G_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(prefix, g, o, i, y, x) \
+    get_g_os_is_yx_isa8_osv8_isv2_index(                    \
+        g, o, i, y, x, CAT(prefix, _SIZE_X ),                        \
+        CAT(prefix, _SIZE_Y),                                        \
+        CAT(prefix, _IFM_NUM),                                       \
+        CAT(prefix, _OFM_NUM),                                       \
+        CAT(prefix, _OFFSET))
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:476]:
```suggestion
#define GET_FILTER_G_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(prefix, g, o, i, y, x) \
    get_g_os_is_yx_isa8_osv8_isv2_index(                                  \
        g, o, i, y, x, CAT(prefix, _SIZE_X ),                             \
        CAT(prefix, _SIZE_Y),                                             \
        CAT(prefix, _IFM_NUM),                                            \
        CAT(prefix, _OFM_NUM),                                            \
        CAT(prefix, _OFFSET))
```

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 9]
```diff
@@ -419,6 +419,78 @@ inline uint get_giy_xs_os_xsv2_osv_index(uint g, uint o, uint i, uint y, uint x,
...
+
+#define GET_FILTER_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(prefix, o, i, y, x) \
+    get_g_os_is_yx_isa8_osv8_isv2_index(                    \
+        0, o, i, y, x, CAT(prefix, _SIZE_X ),                        \
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:480]:
```suggestion
#define GET_FILTER_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(prefix, o, i, y, x) \
    get_g_os_is_yx_isa8_osv8_isv2_index(                             \
        0, o, i, y, x, CAT(prefix, _SIZE_X ),                        \
```

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 10]
```diff
@@ -606,6 +678,34 @@ inline uint get_os_is_zyx_osa4_isa8_osv8_isv4_swizzled_by_4_index(uint o, uint i
     return idx;
 }
 
+inline uint get_g_is_os_yx_osa4_isa8_osv8_isv4(uint g, uint o, uint i, uint z, uint y, uint x,
+                                                     uint size_x, uint size_y, uint size_z, uint size_ifm, uint size_ofm, uint offset)
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:682]: nit: alignment

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 11]
```diff
@@ -606,6 +678,34 @@ inline uint get_os_is_zyx_osa4_isa8_osv8_isv4_swizzled_by_4_index(uint o, uint i
...
+{
+    const uint isv_idx = i % 4;
+    const uint isa_idx = (i / 4) % 8;
+    const uint is_idx = (i / 32);
+    const uint osv_idx = o % 8;
+    const uint osa_idx = (o / 8) % 4;
+    const uint os_idx = (o / 32);
+
+    const uint ifm_32_aligned = ((size_ifm + 31)/32);
+    const uint ofm_32_aligned = ((size_ofm + 31)/32);
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:692]:
```suggestion
    const uint ifm_32_aligned = ((size_ifm + 31) / 32);
    const uint ofm_32_aligned = ((size_ofm + 31) / 32);
```

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl [hunk 12]
```diff
@@ -754,6 +854,16 @@ inline uint get_g_is_os_yx_isa4_osa8_isv8_osv4(uint g, uint o, uint i, uint z, u
     return get_g_os_is_yx_osa4_isa8_osv8_isv4(g, i, o, z, y, x, size_x, size_y, size_z, size_ofm, size_ifm, offset);
 }
 
+#define GET_FILTER_IS_OS_YX_OSA4_ISA8_OSV8_ISV4_INDEX(prefix, o, i, y, x) \
+    get_g_is_os_yx_osa4_isa8_osv8_isv4(                        \
+        0, o, i, 0, y, x,                                                 \
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:859]:
```suggestion
#define GET_FILTER_IS_OS_YX_OSA4_ISA8_OSV8_ISV4_INDEX(prefix, o, i, y, x) \
    get_g_is_os_yx_osa4_isa8_osv8_isv4(                                   \
        0, o, i, 0, y, x,                                                 \
```

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl [hunk 13]
```diff
@@ -71,6 +71,12 @@ inline uint FUNC(get_input_index)(uint g, uint o, uint i, uint z, uint y, uint x
     return GET_FILTER_IS_OS_ZYX_ISV16_OSV16_INDEX(INPUT0, o, i, z, y, x, SUB_GROUP_SIZE);
 #elif defined INPUT0_LAYOUT_IS_OS_YX_ISV16_OSV16
     return GET_FILTER_IS_OS_YX_ISV16_OSV16_INDEX(INPUT0, o, i, y, x, SUB_GROUP_SIZE);
+#elif defined INPUT0_LAYOUT_G_OS_IS_YX_ISA8_OSV8_ISV2
+	return GET_FILTER_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(INPUT0, g, o, i, y, x);
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl:75]: Please change all tabulations on spaces.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl [hunk 14]
```diff
@@ -249,6 +259,12 @@ inline uint FUNC(get_output_index)(uint g, uint o, uint i, uint z, uint y, uint
     return GET_FILTER_IS_OS_ZYX_ISV16_OSV16_INDEX(OUTPUT, o, i, z, y, x, SUB_GROUP_SIZE);
 #elif defined OUTPUT_LAYOUT_IS_OS_YX_ISV16_OSV16
     return GET_FILTER_IS_OS_YX_ISV16_OSV16_INDEX(OUTPUT, o, i, y, x, SUB_GROUP_SIZE);
+#elif defined OUTPUT_LAYOUT_G_OS_IS_YX_ISA8_OSV8_ISV2
+	return GET_FILTER_G_OS_IS_YX_ISA8_OSV8_ISV2_INDEX(OUTPUT, g, o, i, y, x);
```
Comment (lznamens) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl:263]: Tab -> spaces

**Review by lznamens — CHANGES_REQUESTED** [review]
In general LGTM, please resolve code style notes.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 1 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp]: updated it.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 2 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp]: updated it.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 3 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp]: updated it.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp (hunk 4 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 5 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 6 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl:425]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 8 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 10 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 12 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl (hunk 14 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/reorder_weights.cl]: updated it.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl (hunk 7 above)
Comment (hyunback) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/include/batch_headers/fetch_weights.cl]: updated it.

File: src/plugins/intel_gpu/tests/CMakeLists.txt [hunk 15]
```diff
@@ -34,7 +34,7 @@ endif()
 
 # Workaround to avoid warnings during LTO build
 if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
-    set_target_properties(${TARGET_NAME} PROPERTIES LINK_FLAGS_RELEASE "-Wno-error=maybe-uninitialized -Wno-maybe-uninitialized")
+    set_target_properties(${TARGET_NAME} PROPERTIES LINK_FLAGS_RELEASE "-Wno-error=maybe-uninitialized -Wno-maybe-uninitialized -Wno-odr")
```
Comment (isanghao) [inline src/plugins/intel_gpu/tests/CMakeLists.txt:37]: is `no-odr` required at this moment?

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 16]
```diff
@@ -959,10 +943,15 @@ layout layout_optimizer::get_expected_layout(layout const& current_layout,
         /* ***************************** OneDNN impls format selection part ****************************** */
         bool valid_grouped = !is_dw && prim->groups > 1 && (ofm_per_group % compute_block == 0 && ifm_per_group % compute_block == 0);
         bool i8_u8_output = data_type_traits::is_i8_u8(output_layout.data_type);
-        // bool is_first_conv = input_layout.size.feature[0] < 4;
+        bool fp16_output = output_layout.data_type == data_types::f16;
 
-        if (i8_u8_output) {
-            if ((non_grouped || valid_grouped || valid_int8_dw) && onednn_valid_post_ops && is_2d) {
+        // oneDNNv.26 needs acdb format for shallow group conv
+        if (!non_grouped && is_2d &&
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:949]: This condition is unnecessary, See 932line, already check it.

File: src/plugins/intel_gpu/tests/CMakeLists.txt (hunk 15 above)
Comment (hyunback) [reply src/plugins/intel_gpu/tests/CMakeLists.txt:37]: Agree, will remove it.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 17]
```diff
@@ -959,10 +943,15 @@ layout layout_optimizer::get_expected_layout(layout const& current_layout,
         /* ***************************** OneDNN impls format selection part ****************************** */
         bool valid_grouped = !is_dw && prim->groups > 1 && (ofm_per_group % compute_block == 0 && ifm_per_group % compute_block == 0);
         bool i8_u8_output = data_type_traits::is_i8_u8(output_layout.data_type);
-        // bool is_first_conv = input_layout.size.feature[0] < 4;
+        bool fp16_output = output_layout.data_type == data_types::f16;
 
-        if (i8_u8_output) {
-            if ((non_grouped || valid_grouped || valid_int8_dw) && onednn_valid_post_ops && is_2d) {
+        // oneDNNv.26 needs acdb format for shallow group conv
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:948]: Okey, will update it.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp [hunk 18]
```diff
@@ -584,14 +584,19 @@ void reorder_inputs::run(program& p, layout_optimizer& lo, reorder_factory& rf)
...
-            if (new_layout.data_type == data_types::f16)
-                new_layout.format = format::bs_fs_yx_bsv8_fsv2;
-            else if (data_type_traits::is_i8_u8(new_layout.data_type))
-                new_layout.format = format::bs_fs_yx_bsv8_fsv4;
+            if (new_layout.data_type == data_types::f16) {
+                new_layout.format = (input_layout.batch() < 8) ? format::b_fs_yx_fsv2 : format::bs_fs_yx_bsv8_fsv2;
+            } else if (data_type_traits::is_i8_u8(new_layout.data_type)) {
+                new_layout.format = (input_layout.batch() < 8) ? format::b_fs_yx_fsv4 : format::bs_fs_yx_bsv8_fsv4;
+            }
+            // TODO: handling other types for first conv
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/reorder_inputs.cpp:599]: Maybe no reach any cases, just add TODO if need else statement.

File: src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp [hunk 19]
```diff
@@ -303,18 +388,50 @@ cldnn::format find_format(dnnl::memory::desc desc, bool is_grouped) {
...
+                  [&strides] (size_t ind_l, size_t ind_r) {
+                      return strides[ind_l] > strides[ind_r];
+                  });
+
         if (is_grouped) {
-            throw std::runtime_error(std::string("Unsupported grouped onednn dnnl::memory::desc find_format"));
+            if (desc.data.ndims == 5 && blk.inner_nblks == 3
+                && blk.inner_blks[0] == 8 && blk.inner_blks[1] == 8 && blk.inner_blks[2] == 2
+                && blk.inner_idxs[0] == 2 && blk.inner_idxs[1] == 1 && blk.inner_idxs[2] == 2
+                && order[0] == 0 && order[1] == 1 && order[2] == 2 && order[3] == 3 && order[4] == 4) {
```
[the remark this answers is not in the document]
Comment (hyunback) [reply src/plugins/intel_gpu/src/graph/impls/onednn/utils.cpp:405]: Okey, I will update it.

**Review by isanghao — APPROVED** [review]

**Pull request merged**